### PR TITLE
Bitbucket Server support for user provided campaigns credentials

### PR DESCRIPTION
**Pull request by eseliger**

### CODE REVIEW

**Comment by mrnugget** [comment]
@eseliger ready from my side! Please take a look at the commits I pushed :)

File: enterprise/internal/campaigns/store_codehost.go [hunk 1]
```diff
@@ -55,3 +56,43 @@ func scanCodeHost(c *campaigns.CodeHost, sc scanner) error {
 		&c.ExternalServiceID,
 	)
 }
+
+type GetExternalServiceIDOpts struct {
+	ExternalServiceType string
+	ExternalServiceID   string
+}
+
+func (s *Store) GetExternalServiceID(ctx context.Context, opts GetExternalServiceIDOpts) (int64, error) {
```
Comment (eseliger) [inline enterprise/internal/campaigns/store_codehost.go:65]: I think this needs a test.

**Review by eseliger — COMMENTED** [review]
Thanks for finishing this up! 

Do you think it might be worthwhile adding a test for the `kind: bitbucketserver` case in the `CreateCampaignsCredential` resolver?

File: cmd/repo-updater/repos/sources.go [hunk 2]
```diff
@@ -125,6 +125,14 @@ type ChangesetSource interface {
 	ReopenChangeset(context.Context, *Changeset) error
 }
 
+// A UsernameSource can fetch the username associated with the credentials used by the Source.
```
[the remark this answers is not in the document]
Comment (mrnugget) [reply cmd/repo-updater/repos/sources.go:128]: I needed it so I can inject my `FakeSource` in the tests. We can also move it next to the call site, as an unexported interface?

File: enterprise/internal/campaigns/store_codehost.go (hunk 1 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/store_codehost.go:65]: Hm, it's covered by the test for the method in the service, but okay.

File: cmd/repo-updater/repos/sources.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (eseliger) [reply cmd/repo-updater/repos/sources.go:128]: Hmm yeah maybe? Then it's not a concern to anyone outside of campaigns, at least.

[the remark this answers is not in the document]
Comment (mrnugget) [reply cmd/repo-updater/repos/sources.go]: Moved!

**Comment by chrispine** [comment]
Again, so awesome to see this make it in. Thanks!

File: enterprise/internal/campaigns/service.go [hunk 3]
```diff
@@ -427,3 +429,69 @@ func checkRepoSupported(repo *types.Repo) error {
...
+
+// FetchUsernameForBitbucketServerToken fetches the username associated with a
+// Bitbucket server token.
+//
+// We need the username in order to use the token as the password in a HTTP
+// BasicAuth username/password pair used by gitserver to push commits.
+//
+// In order to not require from users to type in their BitbucketServer username
+// we only ask for a token and then use that token to talk to the
+// BitbucketServer API and get their username.
```
Comment (LawnGnome) [inline enterprise/internal/campaigns/service.go:441]: Oh, that's clever.

File: internal/extsvc/bitbucketserver/client.go [hunk 4]
```diff
@@ -157,6 +136,18 @@ func newClient(config *schema.BitbucketServerConnection, httpClient httpcli.Doer
...
+// with the given authenticator instance.
+func (c *Client) WithAuthenticator(a auth.Authenticator) *Client {
+	return &Client{
+		httpClient: c.httpClient,
+		URL:        c.URL,
+		RateLimit:  c.RateLimit,
+		Auth:       a,
+	}
+}
+
```
Comment (LawnGnome) [inline internal/extsvc/bitbucketserver/client.go:150]: Thanks for migrating this client to this pattern — I came to like it better, but ran out of time to retrofit it back after doing the initial work.

File: internal/extsvc/bitbucketserver/client.go [hunk 5]
```diff
@@ -1356,3 +1359,24 @@ func (e *httpError) ExtractExistingPullRequest() (*PullRequest, error) {
 
 	return nil, errors.New("existing PR not found")
 }
+
+// AuthenticatedUsername returns the username associated with the credentials
+// used by the client.
+// Since BitbucketServer doesn't offer an endpoint in their API to query the
+// currently-authenticated user, we send a request to list a single user on the
+// instance and then inspect the response headers in which BitbucketServer sets
+// the username in X-Ausername.
```
Comment (LawnGnome) [inline internal/extsvc/bitbucketserver/client.go:1368]: I'm _slightly_ afraid of this not being actually documented (at least as far as I can tell with a quick Google), but if it seems reliable, fair enough. It's not like there's an obvious REST endpoint we can hit to get the same information, based on a quick scan of the docs.

**Review by LawnGnome — COMMENTED** [review]
Gave this a more detailed post hoc review than I normally would for something a few days old, but this LGTM. Thanks for implementing this!

File: internal/extsvc/bitbucketserver/client.go (hunk 5 above)
Comment (eseliger) [reply internal/extsvc/bitbucketserver/client.go:1368]: I totally get your point. We've looked around and couldn't find any endpoint that does this reliably, though 😢 only BB cloud has one..

**Review by keegancsmith — COMMENTED** [review]
gitserver changes LGTM

File: internal/extsvc/bitbucketserver/client.go (hunk 4 above)
Comment (mrnugget) [reply internal/extsvc/bitbucketserver/client.go:150]: Happy to hear that! I was unsure what the desired pattern was after looking at `bitbucketserver/client.go` but I went with your last-used pattern in the other PR.

File: internal/extsvc/bitbucketserver/client.go (hunk 5 above)
Comment (mrnugget) [reply internal/extsvc/bitbucketserver/client.go:1368]: Erik had the idea of validating tokens before we use them: https://github.com/sourcegraph/sourcegraph/issues/15692

If we add that as an explicit step for _all tokens_, we can rework this code and rethink whether this request makes sense.

**Pull request merged**